File: go-sdk/ui/schema.go
/*
 *  Copyright (c) 2026 Mikhail Knyazhev <markus621@yandex.com>. All rights reserved.
 *  Use of this source code is governed by a BSD 3-Clause license that can be found in the LICENSE file.
 */

// Package ui defines the transport-independent declarative UI protocol.
package ui

//go:generate easyjson

import (
	"errors"
	"fmt"
	"strings"
)

const (
	// ProtocolVersion is the supported declarative UI protocol version.
	ProtocolVersion = "1.0"
	// UIMIMEType is the MIME type used for declarative UI resources.
	UIMIMEType = "application/vnd.osspkg.ui+json"
)

var (
	// ErrInvalidSchema reports an invalid manifest or view schema.
	ErrInvalidSchema = errors.New("invalid ui schema")
	// ErrInvalidLayout reports an invalid node layout.
	ErrInvalidLayout = errors.New("invalid ui layout")
	// ErrInvalidValue reports an invalid declarative value.
	ErrInvalidValue = errors.New("invalid ui value")
)

// Manifest describes the application entry points exposed to a host.
//
//easyjson:json
//nolint:recvcheck // easyjson generates value marshal and pointer unmarshal methods.
type Manifest struct {
	ProtocolVersion    string             `json:"protocolVersion"`
	Plugin             PluginManifest     `json:"plugin"`
	Views              []UIViewDescriptor `json:"views"`
	RequiredComponents []string           `json:"requiredComponents,omitempty"`
}

// PluginManifest identifies the plugin that owns the UI.
//
//easyjson:json
//nolint:recvcheck // easyjson generates value marshal and pointer unmarshal methods.
type PluginManifest struct {
	ID          string `json:"id"`
	Title       string `json:"title"`
	Version     string `json:"version"`
	Description string `json:"description,omitempty"`
}

// UIViewDescriptor identifies a view schema resource in a manifest.
//
//easyjson:json
//nolint:recvcheck // easyjson generates value marshal and pointer unmarshal methods.
type UIViewDescriptor struct {
	ID     string `json:"id"`
	Title  string `json:"title"`
	Route  string `json:"route,omitempty"`
	Schema string `json:"schema"`
}

// ViewSchema is the declarative layout, state, source, and action model for a view.
//
//easyjson:json
//nolint:recvcheck // easyjson generates value marshal and pointer unmarshal methods.
type ViewSchema struct {
	ProtocolVersion string                `json:"protocolVersion"`
	ID              string                `json:"id"`
	Revision        string                `json:"revision,omitempty"`
	Title           string                `json:"title,omitempty"`
	State           map[string]any        `json:"state,omitempty"`
	Sources         map[string]DataSource `json:"sources,omitempty"`
	Actions         map[string]Action     `json:"actions,omitempty"`
	Regions         Regions               `json:"regions"`
}

// Regions groups view nodes into the host-supported layout regions.
//
//easyjson:json
//nolint:recvcheck // easyjson generates value marshal and pointer unmarshal methods.
type Regions struct {
	TopHeader  []Node `json:"top-header"`
	LeftPanel  []Node `json:"left-panel"`
	Content    []Node `json:"content"`
	RightPanel []Node `json:"right-panel"`
	Bottom     []Node `json:"bottom"`
}

// Layout places a node in a twelve-column row layout.
//
//easyjson:json
//nolint:recvcheck // easyjson generates value marshal and pointer unmarshal methods.
type Layout struct {
	Row    int `json:"row"`
	Cols   int `json:"cols,omitempty"`
	Offset int `json:"offset,omitempty"`
}

// Node describes one declarative UI component and its descendants.
//
//easyjson:json
//nolint:recvcheck // easyjson generates value marshal and pointer unmarshal methods.
type Node struct {
	ID        string                  `json:"id"`
	Component string                  `json:"component"`
	Layout    *Layout                 `json:"layout,omitempty"`
	Props     map[string]Value        `json:"props,omitempty"`
	Events    map[string]EventHandler `json:"events,omitempty"`
	Children  []Node                  `json:"children,omitempty"`
	Slots     map[string][]Node       `json:"slots,omitempty"`
	When      *Value                  `json:"when,omitempty"`
}

// EventHandler maps a UI event to an action or sequence of action steps.
//
//easyjson:json
//nolint:recvcheck // easyjson generates value marshal and pointer unmarshal methods.
type EventHandler struct {
	Action string       `json:"action,omitempty"`
	Steps  []ActionStep `json:"steps,omitempty"`
}

// DataSource describes a host tool used to populate view data.
//
//easyjson:json
//nolint:recvcheck // easyjson generates value marshal and pointer unmarshal methods.
type DataSource struct {
	Type      string           `json:"type"`
	Tool      string           `json:"tool"`
	Input     map[string]Value `json:"input,omitempty"`
	Policy    SourcePolicy     `json:"policy,omitempty"`
	Cache     *CachePolicy     `json:"cache,omitempty"`
	RefreshOn []string         `json:"refreshOn,omitempty"`
}

// SourcePolicy controls when a data source is fetched.
type SourcePolicy string

const (
	// SourceManual fetches the source only when explicitly requested.
	SourceManual SourcePolicy = "manual"
	// SourceOnMount fetches the source when the view is mounted.
	SourceOnMount SourcePolicy = "on-mount"
)

// CachePolicy controls how long fetched source data may be reused.
//
//easyjson:json
//nolint:recvcheck // easyjson generates value marshal and pointer unmarshal methods.
type CachePolicy struct {
	TTL int64 `json:"ttl,omitempty"`
}

// Action describes a named operation available to the view.
//
//easyjson:json
//nolint:recvcheck // easyjson generates value marshal and pointer unmarshal methods.
type Action struct {
	Type    string           `json:"type"`
	Tool    string           `json:"tool,omitempty"`
	Input   map[string]Value `json:"input,omitempty"`
	Path    string           `json:"path,omitempty"`
	Value   *Value           `json:"value,omitempty"`
	Source  string           `json:"source,omitempty"`
	To      string           `json:"to,omitempty"`
	Variant string           `json:"variant,omitempty"`
	Message string           `json:"message,omitempty"`
	Effects []Effect         `json:"effects,omitempty"`
}

// ActionStep is one step in an event action sequence.
//
//easyjson:json
//nolint:recvcheck // easyjson generates value marshal and pointer unmarshal methods.
type ActionStep struct {
	Type   string           `json:"type"`
	Action string           `json:"action,omitempty"`
	Input  map[string]Value `json:"input,omitempty"`
}

// Effect describes a client-side consequence of an action.
//
//easyjson:json
//nolint:recvcheck // easyjson generates value marshal and pointer unmarshal methods.
type Effect struct {
	Type         string `json:"type"`
	Path         string `json:"path,omitempty"`
	Source       string `json:"source,omitempty"`
	Value        *Value `json:"value,omitempty"`
	BaseRevision string `json:"baseRevision,omitempty"`
	Revision     string `json:"revision,omitempty"`
	To           string `json:"to,omitempty"`
	Variant      string `json:"variant,omitempty"`
	Message      string `json:"message,omitempty"`
}

// Validate checks whether the manifest contains valid plugin and view entries.
func (m Manifest) Validate() error {
	if m.ProtocolVersion != ProtocolVersion {
		return fmt.Errorf("%w: unsupported manifest protocol version %q", ErrInvalidSchema, m.ProtocolVersion)
	}

	if m.Plugin.ID == "" {
		return fmt.Errorf("%w: plugin id is required", ErrInvalidSchema)
	}

	seen := make(map[string]struct{}, len(m.Views))
	for _, view := range m.Views {
		if view.ID == "" || view.Schema == "" {
			return fmt.Errorf("%w: view id and schema are required", ErrInvalidSchema)
		}

		if _, ok := seen[view.ID]; ok {
			return fmt.Errorf("%w: duplicate view %q", ErrInvalidSchema, view.ID)
		}

		seen[view.ID] = struct{}{}
	}

	return nil
}

func validName(name string) bool {
	return name != "" && !strings.ContainsAny(name, "\r\n")
}

File: react-sdk/ui-core/src/schema.ts
export const protocolVersion = "1.0" as const;

export type RegionName = "top-header" | "left-panel" | "content" | "right-panel" | "bottom";
export type SourcePolicy = "manual" | "on-mount";
export type ValueReference = "$state" | "$context" | "$source" | "$event" | "$result";
export type ExpressionOperator = "$eq" | "$ne" | "$gt" | "$gte" | "$lt" | "$lte" | "$and" | "$or" | "$not" | "$in" | "$exists" | "$concat" | "$coalesce" | "$if";
export type ActionType = "tool" | "set-state" | "merge-state" | "refresh-source" | "invalidate";
export type EffectType = "set-state" | "merge-state" | "invalidate" | "refresh-source" | "refresh-view" | "patch-view" | "navigate" | "toast" | "dialog" | "close-dialog";

export interface Manifest {
  protocolVersion: string;
  plugin: { id: string; title: string; version: string; description?: string };
  views: ViewDescriptor[];
  requiredComponents?: string[];
}

export interface ViewDescriptor {
  id: string;
  title: string;
  route?: string;
  schema: string;
}

export interface ViewSchema {
  protocolVersion: string;
  id: string;
  revision?: string;
  title?: string;
  state?: Record<string, unknown>;
  sources?: Record<string, DataSource>;
  actions?: Record<string, ActionDefinition>;
  regions: Record<RegionName, UINode[]>;
}

export interface GridLayout {
  row: number;
  cols?: number;
  offset?: number;
}

export interface UINode {
  id: string;
  component: string;
  layout?: GridLayout;
  props?: Record<string, UIValue>;
  events?: Record<string, EventHandler>;
  children?: UINode[];
  slots?: Record<string, UINode[]>;
  when?: UIValue;
}

export interface DataSource {
  type: "tool";
  tool: string;
  input?: Record<string, UIValue>;
  policy?: SourcePolicy;
  cache?: { ttl?: number };
  refreshOn?: string[];
}

export interface EventHandler {
  action?: string;
  steps?: ActionStep[];
}

export interface ActionStep {
  type: "action";
  action?: string;
  input?: Record<string, UIValue>;
}

export interface ActionDefinition {
  type: ActionType;
  tool?: string;
  input?: Record<string, UIValue>;
  path?: string;
  value?: UIValue;
  source?: string;
  to?: string;
  variant?: string;
  message?: string;
  effects?: Effect[];
}

export interface Effect {
  type: EffectType;
  path?: string;
  source?: string;
  value?: UIValue;
  baseRevision?: string;
  revision?: string;
  to?: string;
  variant?: string;
  message?: string;
}

export type JSONValue = string | number | boolean | null | readonly JSONValue[] | { [key: string]: JSONValue };
export type UIReference = { $state: string } | { $context: string } | { $source: string } | { $event: string } | { $result: string };
export type UIExpression = { [operator in ExpressionOperator]?: readonly UIValue[] };
export type UIValue = JSONValue | UIReference | UIExpression;

export interface ResolveScope {
  state: unknown;
  context: unknown;
  sources: Record<string, SourceRuntimeState>;
  event?: unknown;
  result?: unknown;
}

export interface SourceRuntimeState<T = unknown> {
  status: "idle" | "loading" | "success" | "error";
  data?: T;
  error?: RuntimeError;
  updatedAt?: number;
  requestId?: string;
}

export interface RuntimeError {
  code: string;
  message: string;
  details?: unknown;
  retryable?: boolean;
}

export const regionNames: readonly RegionName[] = ["top-header", "left-panel", "content", "right-panel", "bottom"];
